### PR TITLE
Added several params to training model

### PR DESCRIPTION
**Pull request by CatEek**
### Description

Added several params for lightning trainer into respective pydantic model

### Changes Made

- **Added**: Some features for the trainer class. More advanced features should only be used in lightning api

**Please ensure your PR meets the following requirements:**

- [x] Code builds and passes tests locally, including doctests
- [x] New tests have been added (for bug fixes/features)
- [x] Pre-commit passes
- [ ] PR to the documentation exists (for bug fixes / features)

### CODE REVIEW

**Comment by jdeschamps** [comment]
One of the test if failing!

File: src/careamics/config/training_model.py [hunk 1]
```diff
@@ -37,6 +37,20 @@ class TrainingConfig(BaseModel):
     num_epochs: int = Field(default=20, ge=1)
     """Number of epochs, greater than 0."""
 
+    precision: Literal["64", "32", "16-mixed", "bf16-mixed"] = Field(default="32")
+    """Numerical precision"""
+    max_steps: int = Field(default=-1, ge=-1, exclude=0)
```
Comment (jdeschamps) [inline src/careamics/config/training_model.py]: `exclude` excludes the field from model dumps (see [pydantic notes](https://docs.pydantic.dev/latest/concepts/fields/#exclude)), it does not exclude values.

You will need a field validator here.

```python
from pydantic import BaseModel, Field

class Stuff(BaseModel):
    max_steps: int = Field(default=-1, ge=-1, exclude=0)

Stuff(max_steps=0)
```

File: src/careamics/config/training_model.py [hunk 2]
```diff
@@ -37,6 +37,20 @@ class TrainingConfig(BaseModel):
...
+    """Maximum number of steps to train for. -1 means no limit."""
+    check_val_every_n_epoch: Optional[int] = Field(default=1, ge=1)
+    """Validation step frequency."""
+    enable_progress_bar: Optional[bool] = Field(default=True)
```
Comment (jdeschamps) [inline src/careamics/config/training_model.py]: Why are `check_val_every_n_epoch` and `enable_progress_bar` optional? Does setting `None` has a meaning here?

I see that they are so in the PL Trainer, but in our case since they have a default value, they will never be `None`.

File: src/careamics/config/training_model.py [hunk 3]
```diff
@@ -37,6 +37,20 @@ class TrainingConfig(BaseModel):
...
+    """Whether to enable the progress bar."""
+    accumulate_grad_batches: int = Field(default=1, ge=1)
+    """Number of batches to accumulate gradients over before stepping the optimizer."""
+    gradient_clip_val: Optional[Union[int, float]] = None
+    """The value to which to clip the gradient"""
+    gradient_clip_algorithm: Optional[Literal["value", "norm"]] = "norm"
```
Comment (jdeschamps) [inline src/careamics/config/training_model.py]: Same comment here, will `gradient_clip_algorithm` ever be set to `None`?

File: tests/config/test_training_model.py [hunk 4]
```diff
@@ -11,3 +11,11 @@ def test_training_wrong_values_by_assignments(minimum_training: dict):
     training.num_epochs = 2
     with pytest.raises(ValueError):
         training.num_epochs = -1
+
+    training.max_steps = 2
+    with pytest.raises(ValueError):
+        training.max_steps = 0
```
Comment (jdeschamps) [inline tests/config/test_training_model.py:17]: This definitely will not pass. That might be the test that is currently failing in the PR!

**Review by jdeschamps — CHANGES_REQUESTED** [review]

File: src/careamics/config/training_model.py
Comment (federico-carrara) [inline src/careamics/config/training_model.py:1]: The overall structure of this works for me :)

File: examples/example_training_LVAE_split.ipynb
Comment (federico-carrara) [inline examples/example_training_LVAE_split.ipynb:1]: Before merging this PR, please revert the changes here in the notebook (otherwise it would be a mess with conflicts...)

File: src/careamics/models/lvae/likelihoods.py [hunk 5]
```diff
@@ -355,6 +356,8 @@ def log_likelihood(self, x: torch.Tensor, params: dict[str, torch.Tensor]):
         torch.Tensor
             The log-likelihood tensor. Shape is (B, C, [Z], Y, X).
         """
+        self.data_mean = self.data_mean.to(x.device)
+        self.data_std = self.data_std.to(x.device)  # TODO this is ugly fix !!
```
Comment (federico-carrara) [inline src/careamics/models/lvae/likelihoods.py]: This can stay temporarily, but would be changed soon since it is not very clean in my opinion.

Comment (CatEek) [reply src/careamics/models/lvae/likelihoods.py]: This comes from previous pr and we discussed it should change

**Review by federico-carrara — APPROVED** [review]

File: src/careamics/config/training_model.py [hunk 6]
```diff
@@ -37,15 +37,15 @@ class TrainingConfig(BaseModel):
...
-    check_val_every_n_epoch: Optional[int] = Field(default=1, ge=1)
+    check_val_every_n_epoch: int = Field(default=1, ge=1)
     """Validation step frequency."""
-    enable_progress_bar: Optional[bool] = Field(default=True)
+    enable_progress_bar: bool = Field(default=True)
     """Whether to enable the progress bar."""
     accumulate_grad_batches: int = Field(default=1, ge=1)
     """Number of batches to accumulate gradients over before stepping the optimizer."""
-    gradient_clip_val: Optional[Union[int, float]] = None
+    gradient_clip_val: Union[int, float] = None
```
Comment (jdeschamps) [inline src/careamics/config/training_model.py]: But this one is default to `None`, so it is "Optional". Or a `Union[int, float, None]` if `None` has a particular meaning!

**Review by jdeschamps — CHANGES_REQUESTED** [review]

**Comment by jdeschamps** [comment]
There are three pre-commit errors that should be fixed.

**Comment by jdeschamps** [comment]
There are conflics that need to be resolved!

**Comment by jdeschamps** [comment]
```
/pc/clone/2gZISA6fREim0cU7Ekhohw/py_env-python3/lib/python3.12/site-packages/numpydoc/docscrape.py:421: UserWarning: Unknown section Example
  self[section] = content
+-------------------------------------------+--------------------------------------------------+---------+-----------------------------------------+
| file                                      | item                                             | check   | description                             |
+===========================================+==================================================+=========+=========================================+
| src/careamics/config/training_model.py:86 | training_model.TrainingConfig.validate_max_steps | PR01    | Parameters {'max_steps'} not documented |
+-------------------------------------------+--------------------------------------------------+---------+-----------------------------------------+
| src/careamics/config/training_model.py:86 | training_model.TrainingConfig.validate_max_steps | RT01    | No Returns section found                |
+-------------------------------------------+--------------------------------------------------+---------+-----------------------------------------+
```

then we can merge!

**Review by jdeschamps — APPROVED** [review]
